File: scripts/learn-from-fable/transcript_parity.py
#!/usr/bin/env python3
"""Python side of the transcript parity harness (see transcript-parity.ts)."""
import json, os, sys

# SkillOpt is the Python reference implementation this harness compares against.
# Point SKILLOPT_PATH at your checkout; nothing here assumes one machine's layout.
skillopt = os.environ.get("SKILLOPT_PATH")
if not skillopt:
    sys.exit("SKILLOPT_PATH is not set — point it at your SkillOpt checkout to run the parity harness.")

sys.path.insert(0, skillopt)
from skillopt.envs.fable_clone.transcript import load_turns, condense_for_extraction

out = {}
for f in sys.argv[1:]:
    turns = load_turns(f)
    out[f.rsplit("/", 1)[-1]] = {
        "turns": len(turns),
        "assistant": sum(t["role"] == "assistant" for t in turns),
        "user": sum(t["role"] == "user" for t in turns),
        "toolResult": sum(t["role"] == "tool_result" for t in turns),
        "fable": sum(t["role"] == "assistant" and t["is_fable"] for t in turns),
        "sidechain": sum(bool(t.get("is_sidechain")) for t in turns),
        "thinkingChars": sum(len(t["thinking"]) for t in turns),
        "toolCalls": sum(len(t["tools"]) for t in turns),
        "windows": len(condense_for_extraction(turns)),
    }
print(json.dumps(out, indent=2))

File: scripts/learn-from-fable/transcript-parity.ts
/**
 * Parity harness: compares src/learn-from-fable/lib/transcript.ts against
 * SkillOpt's fable_clone/transcript.py on the same session files. Run whenever
 * either side changes. Python side: scripts/learn-from-fable/transcript_parity.py
 * emits the same JSON shape; diff the two outputs.
 *
 * Usage: bun scripts/learn-from-fable/transcript-parity.ts <session.jsonl...>
 */
import { SafeJSON } from "@genesiscz/utils/json";
import { condenseForExtraction, loadTurns } from "../../src/learn-from-fable/lib/transcript";

const files = process.argv.slice(2);
if (!files.length) {
    console.error("usage: bun scripts/learn-from-fable/transcript-parity.ts <session.jsonl...>");
    process.exit(1);
}

const out: Record<string, unknown> = {};
for (const file of files) {
    // dedupeUuids off: the Python reference has no uuid dedupe; parity compares raw parsing only
    const turns = await loadTurns(file, { dedupeUuids: false });
    const windows = condenseForExtraction(turns);
    out[file.split("/").at(-1) ?? file] = {
        turns: turns.length,
        assistant: turns.filter((t) => t.role === "assistant").length,
        user: turns.filter((t) => t.role === "user").length,
        toolResult: turns.filter((t) => t.role === "tool_result").length,
        fable: turns.filter((t) => t.role === "assistant" && t.isFable).length,
        sidechain: turns.filter((t) => t.isSidechain).length,
        thinkingChars: turns.reduce((n, t) => n + t.thinking.length, 0),
        toolCalls: turns.reduce((n, t) => n + t.tools.length, 0),
        windows: windows.length,
    };
}

console.log(SafeJSON.stringify(out, null, 2));
